include/ftxui/component/event_input_listener.hpp

```diff
@@ -0,0 +1,55 @@
+#ifndef FTXUI_COMPONENT_EVENT_INPUT_LISTENER_HPP
+#define FTXUI_COMPONENT_EVENT_INPUT_LISTENER_HPP
+
+#include <atomic>
+#include <functional>
+#include <thread>
+
+#include "event.hpp"
+
+#ifdef WIN32
+#include <condition_variable>
+#include <mutex>
+#include <deque>
+#endif
+
+namespace ftxui {
+
+// Receives input events from the OS and turns them into
+// Event objects and sends to a consumer
+//
+// On NIX systems:
+//   - uses SIGWINCH for resize
+//   - uses getchar() for keypresses
+//
+// On Windows systems:
+//   - Uses ReadConsoleInput for resize and keypresses
+class EventInputListener {
```

Comment:
Why do we need this class?

Can't we spawn some thread listening for event and use PostEvent for synchronization with the main thread?

Reply:
I started out by having just a single thread and a lambda in `ScreenInteractive` but it just became too crowded, also especially since we actually need two threads.

The two threads are required in order to be able to reuse `Event::GetEvent` for input parsing. I was quickly considering turning the parser used by `Event::GetEvent` into a push-parser rather than pull-parser but my pull-request started becoming too big.

Thread 1: is used for waiting and dequeuing `INPUT_RECORD`s using `ReadConsoleInput`, all high-level events are immediate posted (via `consumer`), all characters are put on an internal character stream queue.

Thread 2: (this thread is also used on *NIX)
  - on NIX: uses `::getchar()`
  - on Win: pulls (blockingly) chars from the character stream queue mentioned above

Reply:
> I have issues compiling on Linux after you patch:
> 
> ```c++
> /home/sonzogna/programmation/real/FTXUI/src/ftxui/component/event_input_listener.cpp:158:1: error: expected unqualified-id before ‘while’
>   158 | while (!quit_) {
>       | ^~~~~
> /home/sonzogna/programmation/real/FTXUI/src/ftxui/component/event_input_listener.cpp:164:1: error: expected declaration before ‘}’ token
>   164 | }  // namespace ftxui
>       | ^
> make[2]: *** [CMakeFiles/component.dir/build.make:115: CMakeFiles/component.dir/src/ftxui/component/event_input_listener.cpp.o] Error 1
> make[2]: *** Attente des tâches non terminées....
> ```

fixed in latest commit